vcd/resource_vcd_ip_space.go

```diff
@@ -0,0 +1,427 @@
+	"github.com/vmware/go-vcloud-director/v2/govcd"
+	"github.com/vmware/go-vcloud-director/v2/types/v56"
+)
+
+var ipSpaceIpRangeRange = &schema.Resource{
+	Schema: map[string]*schema.Schema{
+		"start_address": {
+			Type:         schema.TypeString,
+			Required:     true,
+			Description:  "Start address of the IP range",
+			ValidateFunc: validation.IsIPAddress,
+		},
+		"end_address": {
+			Type:         schema.TypeString,
+			Required:     true,
+			Description:  "End address of the IP range",
+			ValidateFunc: validation.IsIPAddress,
+		},
+	},
+}
+
+var ipPrefixes = &schema.Resource{
+	Schema: map[string]*schema.Schema{
+		"prefix": {
+			Type:        schema.TypeSet,
+			Required:    true,
+			Description: "One or more prefixes",
+			Elem:        ipSpacePrefix,
+		},
+		"default_quota": {
+			Type:         schema.TypeString,
+			Optional:     true,
+			Description:  "Floating IP quota",
+			ValidateFunc: IsIntAndAtLeast(-1), // -1 - unlimited, 0 - no quota
+		},
+	},
+}
+
+var ipSpacePrefix = &schema.Resource{
+	Schema: map[string]*schema.Schema{
+		"first_ip": {
+			Type:        schema.TypeString,
+			Required:    true,
+			Description: "First IP",
+		},
+		"prefix_length": {
+			Type:         schema.TypeString,
+			Required:     true,
+			Description:  "Prefix length",
+			ValidateFunc: IsIntAndAtLeast(0),
+		},
+		"prefix_count": {
+			Type:         schema.TypeString,
+			Required:     true,
+			Description:  "Number of prefixes to define",
+			ValidateFunc: IsIntAndAtLeast(1),
+		},
+	},
+}
+
+func resourceVcdIpSpace() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceVcdIpSpaceCreate,
+		ReadContext:   resourceVcdIpSpaceRead,
+		UpdateContext: resourceVcdIpSpaceUpdate,
+		DeleteContext: resourceVcdIpSpaceDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceVcdIpSpaceImport,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"org_id": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Org ID for 'SHARED' IP spaces",
+			},
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Name of IP space",
+			},
+			"description": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "Description of IP space",
+			},
+			"type": {
+				Type:         schema.TypeString,
+				Required:     true,
+				Description:  "Type of IP space",
+				ValidateFunc: validation.StringInSlice([]string{"PUBLIC", "SHARED_SERVICES", "PRIVATE"}, false)},
+			"internal_scope": {
+				Type:        schema.TypeSet,
+				Required:    true,
+				Description: "A set of internal scope IPs in CIDR format",
+				Elem: &schema.Schema{
+					MinItems: 1,
+					Type:     schema.TypeString,
+				},
+			},
+			"ip_range_quota": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				Description:  "IP ranges quota. '-1' - unlimited, '0' - no quota",
+				ValidateFunc: IsIntAndAtLeast(-1),
```

Comment:
Just to double check, does 'no quota' mean no IPs allocated at all?

Reply:
> Just to double check, does 'no quota' mean no IPs allocated at all?

Yep. It sort of means - I have some definitions within IP Space, but for whatever reason I do not want to give it to anyone. Also to note - this is a "general quota". There will be a separate resource (on other PR) which will allow to override the quota per tenant while this quota applies to each tenant.